### PR TITLE
PHOENIX-6447: Add support for SYSTEM.CHILD_LINK table in systemcatalogwalentryfilter

### PR DESCRIPTION
**Pull request by sandeepvinayak**

### CODE REVIEW

**Comment by sandeepvinayak** [comment]
@gjacoby126 @yanxinyi can you please review this to add system child link table in systemcatalog wal entry filter for replication.

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   5m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.16 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  13m 40s |  4.16 passed  |
| +1 :green_heart: |  compile  |   1m  0s |  4.16 passed  |
| +1 :green_heart: |  checkstyle  |   0m 27s |  4.16 passed  |
| +1 :green_heart: |  javadoc  |   0m 45s |  4.16 passed  |
| +0 :ok: |  spotbugs  |   2m 59s |  phoenix-core in 4.16 has 944 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 41s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  2s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m  2s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 27s |  phoenix-core: The patch generated 19 new + 4 unchanged - 7 fixed = 23 total (was 11)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 44s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   3m 11s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 127m 36s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 11s |  The patch does not generate ASF License warnings.  |
|  |   | 164m 17s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1201 |
| JIRA Issue | PHOENIX-6447 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux b45fec623579 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.16 / 0a79cb6 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/1/testReport/ |
| Max. process+thread count | 5623 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/it/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilterIT.java [hunk 1]
```diff
@@ -90,6 +91,7 @@ public static synchronized void setup() throws Exception {
       catalogTable = PhoenixRuntime.getTable(connection, PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME);
       childLinkTable = PhoenixRuntime.getTable(connection, PhoenixDatabaseMetaData.SYSTEM_CHILD_LINK_NAME);
       walKey = new WALKey(REGION, TableName.valueOf(PhoenixDatabaseMetaData.SYSTEM_CATALOG_NAME), 0, 0, uuid);
+      walKey = new WALKey(REGION, TableName.valueOf(PhoenixDatabaseMetaData.SYSTEM_CHILD_LINK_NAME), 0, 0, uuid);
```
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilterIT.java:94]: Doesn't this make the previous line unnecessary?

File: phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java [hunk 2]
```diff
@@ -58,13 +64,63 @@ public Cell filterCell(final WAL.Entry entry, final Cell cell) {
...
+
+  /**
+   * is the cell for system child link a tenant owned.
+   * @param cell hbase cell
+   * @return true if the cell is tenant owned
+   */
+  private boolean isTenantRowCellSystemChildLink(final Cell cell) {
+    ImmutableBytesWritable key = new ImmutableBytesWritable(
+      cell.getRowArray(), cell.getRowOffset(), cell.getRowLength());
+    //rows in system.catalog that aren't tenant-owned will
```
Comment (ankitjain64) [inline phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java:95]: nit: Remove system.catalog from comments.

File: phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java [hunk 3]
```diff
@@ -58,13 +64,63 @@ public Cell filterCell(final WAL.Entry entry, final Cell cell) {
...
+    // have a leading separator byte
+    boolean isTenantRowCell =
```
Comment (ankitjain64) [inline phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java:97]: Does the schema for system.child_link table the same as that for system.catalog? and does this assumption of leading separator byte holds true there as well? If so we can `isTenantRowCellSystemCatalog` and reuse the same function here.

Comment (sandeepvinayak) [reply phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java:97]: The schema remains the same but the system catalog table is not supposed to have tenant-owned data with empty tenant id anymore. Thus, I believe we need different methods.

File: phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java [hunk 4]
```diff
@@ -58,13 +64,64 @@ public Cell filterCell(final WAL.Entry entry, final Cell cell) {
...
+   */
+  private boolean isTenantRowCellSystemChildLink(final Cell cell) {
+    boolean isTenantRowCell = doesKeyHaveLeadingSeparatorByte(cell);
+
+    ImmutableBytesWritable key = new ImmutableBytesWritable(
+      cell.getRowArray(), cell.getRowOffset(), cell.getRowLength());
+    boolean isChildLinkToTenantView = false;
+    if (!isTenantRowCell) {
+      ImmutableBytesWritable cq = new ImmutableBytesWritable(
+        cell.getQualifierArray(), cell.getQualifierOffset(),
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java:104]: nit: can use CellUtil.matchingQualifier.

File: phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java [hunk 5]
```diff
@@ -58,13 +64,64 @@ public Cell filterCell(final WAL.Entry entry, final Cell cell) {
...
+        cell.getQualifierLength());
+      boolean isChildLink =
+        cq.compareTo(PhoenixDatabaseMetaData.LINK_TYPE_BYTES) == 0;
+      if (isChildLink) {
+        ImmutableBytesWritable columnValue = new ImmutableBytesWritable(
+          cell.getValueArray(), cell.getValueOffset(), cell.getValueLength());
+        if (columnValue.compareTo(CHILD_TABLE_BYTES) == 0) {
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/replication/SystemCatalogWALEntryFilter.java:111]: nit: can use CellUtil.matchingValue

**Review by gjacoby126 — CHANGES_REQUESTED** [review]
Looks good, just a couple nits

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 31s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.16 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  13m 54s |  4.16 passed  |
| +1 :green_heart: |  compile  |   1m  0s |  4.16 passed  |
| +1 :green_heart: |  checkstyle  |   0m 26s |  4.16 passed  |
| +1 :green_heart: |  javadoc  |   0m 44s |  4.16 passed  |
| +0 :ok: |  spotbugs  |   2m 59s |  phoenix-core in 4.16 has 944 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   5m 41s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 58s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 58s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 27s |  phoenix-core: The patch generated 4 new + 4 unchanged - 7 fixed = 8 total (was 11)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 45s |  the patch passed  |
| -1 :x: |  spotbugs  |   3m 12s |  phoenix-core generated 1 new + 944 unchanged - 0 fixed = 945 total (was 944)  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 127m 50s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 11s |  The patch does not generate ASF License warnings.  |
|  |   | 159m 30s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:phoenix-core |
|  |  Unread field:field be static?  At SystemCatalogWALEntryFilter.java:[line 49] |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/4/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1201 |
| JIRA Issue | PHOENIX-6447 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux 1c2d68d8a757 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.16 / 0a79cb6 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/4/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| spotbugs | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/4/artifact/yetus-general-check/output/new-spotbugs-phoenix-core.html |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/4/testReport/ |
| Max. process+thread count | 5643 (vs. ulimit of 30000) |
| modules | C: phoenix-core U: phoenix-core |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1201/4/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by gjacoby126** [comment]
@sandeepvinayak - could you please change your commit message to match the JIRA name (with the JIRA number). Then this looks good to go.

**Review by gjacoby126 — APPROVED** [review]
+1, thanks for the patch @sandeepvinayak

**Pull request merged**